external/mqtt/src/test/scala/org/apache/spark/streaming/mqtt/MQTTStreamSuite.scala

```diff
@@ -17,31 +17,65 @@
 
 package org.apache.spark.streaming.mqtt
 
-import org.scalatest.FunSuite
-
+import org.scalatest.{BeforeAndAfter, FunSuite}
+import org.scalatest.concurrent.Eventually
+import scala.concurrent.duration._
 import org.apache.spark.streaming.{Seconds, StreamingContext}
 import org.apache.spark.storage.StorageLevel
 import org.apache.spark.streaming.dstream.ReceiverInputDStream
+import org.eclipse.paho.client.mqttv3._
+import org.eclipse.paho.client.mqttv3.persist.MqttDefaultFilePersistence
 
-class MQTTStreamSuite extends FunSuite {
-
-  val batchDuration = Seconds(1)
+class MQTTStreamSuite extends FunSuite with Eventually with BeforeAndAfter {
 
+  private val batchDuration = Seconds(1)
   private val master: String = "local[2]"
-
   private val framework: String = this.getClass.getSimpleName
+  private val brokerUrl = "tcp://localhost:1883"
+  private val topic = "def"
+  private var ssc: StreamingContext = _
 
-  test("mqtt input stream") {
-    val ssc = new StreamingContext(master, framework, batchDuration)
-    val brokerUrl = "abc"
-    val topic = "def"
+  before {
+    ssc = new StreamingContext(master, framework, batchDuration)
+  }
+  after {
+    if (ssc != null) {
+      ssc.stop()
+      ssc = null
+    }
+  }
 
-    // tests the API, does not actually test data receiving
-    val test1: ReceiverInputDStream[String] = MQTTUtils.createStream(ssc, brokerUrl, topic)
-    val test2: ReceiverInputDStream[String] =
+  test("mqtt input stream") {
+    val sendMessage = "MQTT demo for spark streaming"
+    publishData(sendMessage)
+    val receiveStream: ReceiverInputDStream[String] =
       MQTTUtils.createStream(ssc, brokerUrl, topic, StorageLevel.MEMORY_AND_DISK_SER_2)
-
-    // TODO: Actually test receiving data
+    var receiveMessage: String = ""
+    receiveStream.foreachRDD { rdd =>
+      receiveMessage = rdd.first
+      receiveMessage
```

Comment:
rdd.first can throw error if there are not records in the rdd. If by chance, the first batch in this program does not have the published data and returns an empty RDD, there are hidden exceptions made which is recipe for failures in the future - flaky tests, etc. As I said earlier, see the KafkaStreamSuite for inspiration. It does a RDD.collect() and puts all the received data into a hashmap (could be a simple buffer/list in your case). Then the eventually verifies the data in the buffer.